<div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); padding: 50px 40px; border-radius: 20px; margin-bottom: 30px; box-shadow: 0 20px 60px rgba(102,126,234,0.3); position: relative; overflow: hidden;">
  <div style="position: absolute; top: -50px; right: -50px; width: 200px; height: 200px; background: rgba(255,255,255,0.05); border-radius: 50%;"></div>
  <div style="position: absolute; bottom: -30px; left: -30px; width: 150px; height: 150px; background: rgba(255,255,255,0.03); border-radius: 50%;"></div>
  <div style="position: relative; z-index: 1;">
    <div style="display: inline-block; background: rgba(255,255,255,0.15); padding: 6px 16px; border-radius: 20px; font-size: 0.85em; color: rgba(255,255,255,0.9); margin-bottom: 16px; backdrop-filter: blur(10px);">📘 CookBook Series</div>
    <h1 style="color: white; font-size: 2.8em; margin: 0 0 12px 0; font-weight: 700; letter-spacing: -0.02em; line-height: 1.2;">Lesson 01 — Токенизатор и адаптер Hugging Face (Qwen3)</h1>
    <p style="color: rgba(255,255,255,0.85); font-size: 1.15em; margin: 0 0 24px 0; line-height: 1.6;">Пошаговый разбор построения адаптера токенизатора для моделей Qwen3 на основе Hugging Face c поддержкой chat template, батч‑кодирования и обратимого декодирования, с тестами и проверенными практиками.</p>
    <div style="display: flex; gap: 20px; flex-wrap: wrap;">
      <span style="color: rgba(255,255,255,0.7); font-size: 0.9em;">📅 2026-03-14</span>
      <span style="color: rgba(255,255,255,0.7); font-size: 0.9em;">⏱ 30 мин</span>
      <span style="color: rgba(255,255,255,0.7); font-size: 0.9em;">👤 Автор: @Verbasik</span>
      <span style="color: rgba(255,255,255,0.7); font-size: 0.9em;">🟡 Средний</span>
    </div>
  </div>
</div>

<div style="background: linear-gradient(to right, #f8f9ff, #ffffff); border: 1px solid #e2e8f0; border-left: 5px solid #667eea; padding: 25px 30px; border-radius: 0 16px 16px 0; margin-bottom: 20px; box-shadow: 0 4px 15px rgba(0,0,0,0.04);">
  <h3 style="color: #667eea; margin: 0 0 16px 0; font-size: 1.3em; font-weight: 700;">📋 Содержание</h3>
  <div style="columns: 2; column-gap: 30px;">
    <ol style="color: #4a5568; line-height: 2.2; margin: 0; padding-left: 20px;">
      <li><a href="#section-vvedenie" style="color: #667eea; text-decoration: none;">Введение</a></li>
      <li><a href="#section-celi-i-dod" style="color: #667eea; text-decoration: none;">Цели и DoD</a></li>
      <li><a href="#section-trebovaniya-i-ustanovka" style="color: #667eea; text-decoration: none;">Требования и установка</a></li>
      <li><a href="#section-bazovyj-adapter-tokenizatora" style="color: #667eea; text-decoration: none;">Базовый адаптер токенизатора</a></li>
      <li><a href="#section-chat-template-i-soobshcheniya" style="color: #667eea; text-decoration: none;">Chat template и сообщения</a></li>
      <li><a href="#section-batch-tokenizaciya" style="color: #667eea; text-decoration: none;">Батч‑токенизация</a></li>
      <li><a href="#section-dekodirovanie-i-striming" style="color: #667eea; text-decoration: none;">Декодирование и стриминг</a></li>
      <li><a href="#section-testirovanie-pytest" style="color: #667eea; text-decoration: none;">Тестирование (pytest)</a></li>
      <li><a href="#section-diagnostika-i-otladka" style="color: #667eea; text-decoration: none;">Диагностика и отладка</a></li>
      <li><a href="#section-primery-ispolzovaniya" style="color: #667eea; text-decoration: none;">Примеры использования</a></li>
      <li><a href="#section-checklist-dod" style="color: #667eea; text-decoration: none;">Чеклист DoD</a></li>
      <li><a href="#section-ssylki-i-resursy" style="color: #667eea; text-decoration: none;">Ссылки и ресурсы</a></li>
      <li><a href="#section-voprosy-dlya-samoproverki" style="color: #667eea; text-decoration: none;">Вопросы для самопроверки</a></li>
    </ol>
  </div>
</div>

<div style="background: #ebf8ff; border-left: 4px solid #63b3ed; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #2b6cb0; margin-bottom: 8px; font-size: 0.95em;">📌 Мета‑информация</div>
  <ul style="margin: 0; padding-left: 20px; line-height: 1.6;">
    <li>⏱ <strong>Время чтения:</strong> ~24 мин</li>
    <li>🎯 <strong>Цель:</strong> разработать надёжный адаптер токенизатора для Qwen3 на основе Hugging Face с поддержкой строк, батчей и чат‑сообщений, гарантируя корректные <code>eos/pad</code> и обратимое декодирование</li>
    <li>📋 <strong>Предпосылки:</strong> Python 3.10+, базовые знания <code>transformers</code> и токенизации</li>
  </ul>
</div>

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-vvedenie" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">📖</span> Введение
</h2>

Токенизация — это преобразование текста (включая диалоговые сообщения) в последовательность идентификаторов словаря модели. Для совместимости с семейством Qwen3 и экосистемой Hugging Face мы создаём тонкий, но строгий адаптер, который:

- Единообразно кодирует строки, батчи строк и чат‑сообщения.
- Гарантирует корректные `pad_token_id` и `eos_token_id` (при отсутствии `pad` использует `eos`).
- Сохраняет тип тензоров как `torch.int32` (память/совместимость с рантаймом).
- Поддерживает `apply_chat_template` и обратимое `decode(encode(x))` на практических кейсах.

<div style="background: #ffffff; border: 1px solid #e2e8f0; border-radius: 14px; padding: 24px 28px; margin: 20px 0; box-shadow: 0 2px 8px rgba(0,0,0,0.04);">
  <div style="display: flex; align-items: center; gap: 10px; margin-bottom: 16px;">
    <div style="width: 32px; height: 32px; background: linear-gradient(135deg, #667eea, #764ba2); border-radius: 8px; display: flex; align-items: center; justify-content: center; font-size: 0.9em;">💡</div>
    <h4 style="margin: 0; color: #2d3748; font-size: 1.15em; font-weight: 700;">Интуиция: зачем нужен адаптер?</h4>
  </div>
  <div style="color: #4a5568; line-height: 1.8; font-size: 0.97em;">
    Представьте универсальную розетку‑переходник: ваш код обращается к единым методам (<code>encode</code>, <code>batch_encode</code>, <code>decode</code>), а внутри адаптер подстраивается под любые совместимые токенизаторы Hugging Face, гарантируя одинаковые типы и поведение.
  </div>
</div>

<div style="background: #faf5ff; border-left: 4px solid #b794f4; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #6b46c1; margin-bottom: 6px; font-size: 0.95em;">📐 Ключевая формула</div>
  <div style="line-height: 1.7;">
    При паддинге до длины <code>L</code> стоимость памяти на батч приблизительно пропорциональна <code>B × L</code>, где <code>B</code> — размер батча. Выбирайте <code>L</code> ровно настолько большим, насколько нужно для % кейсов.
  </div>
</div>

Где:
- $B$ — размер батча;
- $L$ — целевая длина последовательности после паддинга;
- Память для int32: 4 байта на токен.

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-celi-i-dod" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">🎯</span> Цели и DoD
</h2>

Цели:
- Универсальный адаптер поверх Hugging Face токенизаторов (Qwen3‑совместимых).
- Поддержка строк, батчей и чат‑сообщений через `apply_chat_template`.
- Типы тензоров — `torch.int32`; корректные `pad/eos`.
- Тесты на round‑trip (`decode(encode(x))`) и батчи.

<table style="width: 100%; border-collapse: collapse; margin: 20px 0; border-radius: 12px; overflow: hidden; box-shadow: 0 2px 10px rgba(0,0,0,0.06);">
  <thead>
    <tr style="background: linear-gradient(135deg, #667eea, #764ba2);">
      <th style="padding: 14px 18px; text-align: left; color: white; font-weight: 600; font-size: 0.95em;">Задача</th>
      <th style="padding: 14px 18px; text-align: left; color: white; font-weight: 600; font-size: 0.95em;">Критерий готовности</th>
    </tr>
  </thead>
  <tbody>
    <tr style="background: #f8f9ff;">
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Адаптер токенизатора</td>
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Единый API, корректные типы и токены</td>
    </tr>
    <tr style="background: #ffffff;">
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Батч‑кодирование</td>
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Списки строк/чатов обрабатываются стабильно</td>
    </tr>
    <tr style="background: #f8f9ff;">
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Декодирование</td>
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Round‑trip проходит на базовых примерах</td>
    </tr>
  </tbody>
</table>

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-trebovaniya-i-ustanovka" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">🧰</span> Требования и установка
</h2>

- Python 3.10+
- Библиотеки: `transformers`, `huggingface_hub`, `torch`, `pytest` (для тестов)
- (Опц.) `accelerate` для ускорения загрузки крупных токенизаторов

<div style="background: #fffbeb; border-left: 4px solid #f59e0b; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #b7791f; margin-bottom: 6px; font-size: 0.95em;">💡 Совет</div>
  <div style="line-height: 1.7;">Для быстрых проверок используйте компактные модели (например, <code>Qwen/Qwen3-0.6B</code>) — семантика токенизатора сохраняется, а время установки и памяти меньше.</div>
</div>

<div style="background: #1a202c; color: #a0aec0; padding: 8px 16px; border-radius: 10px 10px 0 0; margin-bottom: -15px; font-family: 'Fira Code', monospace; font-size: 0.8em; display: flex; justify-content: space-between; align-items: center;">
  <span>🐚 install-deps</span>
  <span style="background: #2d3748; padding: 2px 10px; border-radius: 10px; font-size: 0.85em;">Bash</span>
</div>

In [5]:
# Выполните вручную при необходимости установки зависимостей
!pip -q install "transformers>=4.40.0" "huggingface_hub>=0.21.0" "torch>=2.2.0" "pytest>=8.0.0" "accelerate>=0.28.0"


<div style="background: #1a202c; color: #a0aec0; padding: 8px 16px; border-radius: 10px 10px 0 0; margin-bottom: -15px; font-family: 'Fira Code', monospace; font-size: 0.8em; display: flex; justify-content: space-between; align-items: center;">
  <span>🐍 nano-infer/python/minisgl/utils/hf.py</span>
  <span style="background: #2d3748; padding: 2px 10px; border-radius: 10px; font-size: 0.85em;">Python</span>
</div>

In [4]:
# ── Стандартная библиотека ──
from typing import Any, Dict, List, Literal, Optional, Tuple, Union

# ── Сторонние библиотеки: ML/данные ──
import torch                              # Работа с тензорами
from transformers import AutoTokenizer    # Загрузка токенизаторов HF

# Константы проекта
DEFAULT_MODEL_ID: str = "Qwen/Qwen3-8B"   # Идентификатор модели по умолчанию
DEFAULT_DTYPE: torch.dtype = torch.int32  # Единый тип тензоров для совместимости

# Пояснение: объединяем импорты в кластеры и задаём осмысленные константы.
# Это упрощает поддержку и соблюдает единый стиль в проекте.


<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-bazovyj-adapter-tokenizatora" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">📝</span> Базовый адаптер токенизатора
</h2>

<div style="background: #ebf8ff; border-left: 4px solid #63b3ed; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #2b6cb0; margin-bottom: 6px; font-size: 0.95em;">📌 Определение</div>
  <div style="line-height: 1.7;">Адаптер — класс, выравнивающий интерфейс токенизатора: <code>encode</code>, <code>batch_encode</code>, <code>decode</code>, <code>apply_chat_template</code>, а также обеспечивающий корректные <code>pad/eos</code> и типы тензоров.</div>
</div>

<div style="background: #1a202c; color: #a0aec0; padding: 8px 16px; border-radius: 10px 10px 0 0; margin-bottom: -15px; font-family: 'Fira Code', monospace; font-size: 0.8em; display: flex; justify-content: space-between; align-items: center;">
  <span>
    🐍 derived from:<br>
    - nano-infer/python/infer/utils/hf.py;<br>
    - nano-infer/python/infer/tokenizer/tokenize.py;<br>
    - nano-infer/python/infer/tokenizer/detokenize.py
  </span>
  <span style="background: #2d3748; padding: 2px 10px; border-radius: 10px; font-size: 0.85em; white-space: nowrap;">Python</span>
</div>

In [26]:
class TokenizerAdapter:
    """
    Description:
    ---------------
        Унифицированный адаптер для токенизаторов Hugging Face
        (например, для семейства Qwen3). Обеспечивает единый API для
        кодирования строк, батчей и чат‑сообщений, а также обратимое
        декодирование и корректные типы/токены (pad/eos).

    Args:
    ---------------
        model_id: Идентификатор модели/токенизатора в HF Hub.
        use_fast: Использовать fast‑реализацию (при наличии).
        trust_remote_code: Доверять кастомному коду из репозитория.
        add_eos_token: Принудительно добавлять EOS при кодировании.

    Returns:
    ---------------
        Экземпляр адаптера с инициализированным токенизатором.

    Raises:
    ---------------
        ValueError: Если не удаётся корректно установить pad/eos токены.
        OSError: Если загрузка токенизатора с HF завершилась ошибкой.

    Examples:
    ---------------
        >>> adapter = TokenizerAdapter("Qwen/Qwen3-8B")
        >>> ids = adapter.encode("Привет")
        >>> text = adapter.decode(ids)
        >>> assert "Привет" in text
    """

    def __init__(
        self,
        model_id: str = DEFAULT_MODEL_ID,
        use_fast: bool = True,
        trust_remote_code: bool = True,
        add_eos_token: bool = False,
    ) -> None:
        
        # Сохраняем параметры и флаги
        self.model_id: str = model_id             # Идентификатор модели
        self.add_eos_token: bool = add_eos_token  # Флаг добавления EOS

        try:
            # Пытаемся загрузить токенизатор с HF
            self.tokenizer = AutoTokenizer.from_pretrained(
                model_id,
                use_fast=use_fast,
                trust_remote_code=trust_remote_code,
            )
        except OSError as e:
            # Преобразуем исключение в информативную ошибку
            raise OSError(
                f"Не удалось загрузить токенизатор: {model_id}. Ошибка: {e}"
            ) from e

        # Читаем EOS из токенизатора (если определён)
        eos_id: Optional[int] = getattr(self.tokenizer, "eos_token_id", None)
        # Читаем PAD из токенизатора (если определён)
        pad_id: Optional[int] = getattr(self.tokenizer, "pad_token_id", None)

        # Если нет pad — используем eos в качестве pad (распространённая практика)
        if pad_id is None:
            # Если eos не определён — это критично для совместимости
            if eos_id is None:
                raise ValueError(
                    "Отсутствуют и pad_token_id, и eos_token_id — задайте хотя бы EOS"
                )
            # Выставляем pad равным eos
            self.tokenizer.pad_token = self.tokenizer.eos_token  # type: ignore
            pad_id = self.tokenizer.pad_token_id                 # type: ignore

        # Сохраняем идентификаторы токенов как атрибуты экземпляра
        self.eos_token_id: Optional[int] = eos_id  # EOS может быть None
        self.pad_token_id: int = int(pad_id)       # PAD теперь точно определён

        print(f"[TokenizerAdapter] Итог: eos_token_id={self.eos_token_id}, pad_token_id={self.pad_token_id}")
        print(f"[TokenizerAdapter] Размер словаря: {self.tokenizer.vocab_size:,} токенов")

    def encode(
        self,
        text: str,
        *,
        add_special_tokens: bool = True,
        max_length: Optional[int] = None,
        padding: bool = False,
        truncation: bool = False,
        return_tensors: Literal["pt"] = "pt",
        dtype: torch.dtype = DEFAULT_DTYPE,
    ) -> torch.Tensor:
        """
        Description:
        ---------------
            Кодирует одиночную строку в тензор индексов токенов.

        Args:
        ---------------
            text: Исходная текстовая строка для кодирования.
            add_special_tokens: Добавлять спецтокены токенизатора.
            max_length: Целевая длина (для паддинга/усечения).
            padding: Применять паддинг до max_length.
            truncation: Применять усечение до max_length.
            return_tensors: Формат возврата (поддерживается 'pt').
            dtype: Тип данных тензора (по умолчанию int32).

        Returns:
        ---------------
            torch.Tensor: Тензор индексов токенов формы (L,).

        Raises:
        ---------------
            ValueError: Если запрошен неподдерживаемый return_tensors.

        Examples:
        ---------------
            >>> adapter = TokenizerAdapter()
            >>> ids = adapter.encode("Привет")
            >>> ids.dtype == DEFAULT_DTYPE
            True
        """
        # Проверяем поддерживаемый формат возврата
        if return_tensors != "pt":
            raise ValueError("Поддерживается только return_tensors='pt'")

        # Подготавливаем параметры паддинга
        pad_mode: Union[bool, str] = False      # По умолчанию без паддинга
        if padding and max_length is not None:
            pad_mode = "max_length"             # Паддинг до max_length

        # Вызываем токенизатор HF
        encoded: Dict[str, torch.Tensor] = self.tokenizer(
            text,
            add_special_tokens=add_special_tokens or self.add_eos_token,
            padding=pad_mode,
            truncation=truncation,
            max_length=max_length,
            return_tensors="pt",
        )

        # Достаём тензор индексов и приводим тип к целевому
        input_ids: torch.Tensor = encoded["input_ids"].to(dtype=dtype)
        input_ids = input_ids.squeeze(0)

        print(f"[encode] Результат:  {input_ids.shape} токенов, dtype={input_ids.dtype}")
        print(f"[encode] ID токенов: {input_ids.tolist()}")
        print(f"[encode] Токены:     {self.tokenizer.convert_ids_to_tokens(input_ids.tolist())}")
        
        # Удаляем размерность батча (1, L) -> (L,)
        return input_ids.squeeze(0)

    def batch_encode(
        self,
        texts: List[str],
        *,
        add_special_tokens: bool = True,
        max_length: Optional[int] = None,
        padding: bool = True,
        truncation: bool = True,
        return_tensors: Literal["pt"] = "pt",
        dtype: torch.dtype = DEFAULT_DTYPE,
    ) -> Dict[str, torch.Tensor]:
        """
        Description:
        ---------------
            Кодирует список строк в батч тензоров (input_ids,
            attention_mask) с паддингом.

        Args:
        ---------------
            texts: Список строк для кодирования.
            add_special_tokens: Добавлять спецтокены токенизатора.
            max_length: Целевая длина последовательности.
            padding: Включить паддинг (обычно True для батчей).
            truncation: Включить усечение по max_length.
            return_tensors: Формат возврата ('pt').
            dtype: Тип данных выходных тензоров.

        Returns:
        ---------------
            Dict[str, torch.Tensor]: Поля 'input_ids', 'attention_mask'.

        Raises:
        ---------------
            ValueError: Если список пуст или неподдерживаемый формат.

        Examples:
        ---------------
            >>> adapter = TokenizerAdapter()
            >>> batch = adapter.batch_encode(["a", "b"], max_length=8)
            >>> batch["input_ids"].dtype == DEFAULT_DTYPE
            True
        """
        # Базовые проверки входных данных
        if not texts:
            raise ValueError("Список текстов пуст — нечего кодировать")
        if return_tensors != "pt":
            raise ValueError("Поддерживается только return_tensors='pt'")

        # Режим паддинга: для батча по умолчанию включаем
        pad_mode: Union[bool, str] = padding
        if padding and max_length is not None:
            pad_mode = "max_length"  # Жёсткая длина для стабильных батчей

        # Кодируем батч строк
        encoded = self.tokenizer(
            texts,
            add_special_tokens=add_special_tokens or self.add_eos_token,
            padding=pad_mode,
            truncation=truncation,
            max_length=max_length,
            return_tensors="pt",
        )

        # Приводим типы к целевому и возвращаем словарь тензоров
        outputs: Dict[str, torch.Tensor] = {
            k: v.to(dtype=dtype) for k, v in encoded.items()
        }

        ids = outputs["input_ids"]
        mask = outputs.get("attention_mask")
        print(f"[batch_encode] input_ids shape: {ids.shape}, dtype={ids.dtype}")
        
        if mask is not None:
            pad_counts = (mask == 0).sum(dim=1).tolist()
            print(f"[batch_encode] PAD токенов на строку: {pad_counts}")

        return outputs

    def apply_chat_template(
        self,
        messages: List[Dict[str, str]],
        *,
        add_generation_prompt: bool = True,
    ) -> str:
        """
        Description:
        ---------------
            Формирует текстовую подсказку из списка чат‑сообщений,
            используя шаблон токенизатора (ChatML‑подобный формат).

        Args:
        ---------------
            messages: Список словарей вида {"role", "content"}.
            add_generation_prompt: Добавить заготовку для ответа модели.

        Returns:
        ---------------
            str: Отформатированная строка для последующего кодирования.

        Raises:
        ---------------
            ValueError: Если сообщения имеют неверный формат.

        Examples:
        ---------------
            >>> adapter = TokenizerAdapter()
            >>> prompt = adapter.apply_chat_template([
            ...     {"role": "user", "content": "Привет"}
            ... ])
            >>> isinstance(prompt, str)
            True
        """
        # Минимальная валидация формата сообщений
        if not isinstance(messages, list) or not all(
            isinstance(m, dict) and "role" in m and "content" in m
            for m in messages
        ):
            raise ValueError(
                "messages должны быть списком словарей с ключами 'role' и 'content'"
            )

        # Применяем chat template токенизатора: получаем строку
        formatted: str = self.tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=add_generation_prompt,
        )

        print(f"[apply_chat_template] Итоговый промпт ({len(formatted)} символов):")
        print(f"  {formatted!r}")

        return formatted

    def encode_chat(
        self,
        messages: List[Dict[str, str]],
        *,
        max_length: Optional[int] = None,
        padding: bool = False,
        truncation: bool = False,
        dtype: torch.dtype = DEFAULT_DTYPE,
    ) -> torch.Tensor:
        """
        Description:
        ---------------
            Кодирует чат‑сообщения: сначала применяет шаблон, затем
            кодирует итоговую строку в тензор индексов токенов.

        Args:
        ---------------
            messages: Список словарей {"role", "content"}.
            max_length: Целевая длина.
            padding: Включить паддинг.
            truncation: Включить усечение.
            dtype: Тип данных тензора результата.

        Returns:
        ---------------
            torch.Tensor: Последовательность индексов токенов (L,).

        Examples:
        ---------------
            >>> adapter = TokenizerAdapter()
            >>> ids = adapter.encode_chat([
            ...     {"role": "user", "content": "Как дела?"}
            ... ])
            >>> ids.ndim
            1
        """
        # Формируем подсказку по шаблону
        prompt: str = self.apply_chat_template(messages)
        # Кодируем итоговую строку тем же механизмом, что и одиночный текст
        ids = self.encode(
            prompt,
            max_length=max_length,
            padding=padding,
            truncation=truncation,
            dtype=dtype,
        )

        print(f"[encode_chat] Готово: {ids.shape[0]} токенов для {len(messages)} сообщений")
        return ids

    def decode(
        self,
        ids: Union[List[int], torch.Tensor],
        *,
        skip_special_tokens: bool = True,
    ) -> str:
        """
        Description:
        ---------------
            Декодирует последовательность индексов токенов обратно в
            строку, при необходимости пропуская спец‑токены.

        Args:
        ---------------
            ids: Список индексов или тензор токенов.
            skip_special_tokens: Пропускать ли спец‑токены при декоде.

        Returns:
        ---------------
            str: Восстановленная строка текста.

        Examples:
        ---------------
            >>> adapter = TokenizerAdapter()
            >>> s = adapter.decode(adapter.encode("Тест"))
            >>> "Тест" in s
            True
        """
        # Приводим вход к списку целых чисел
        if isinstance(ids, torch.Tensor):
            as_list: List[int] = [int(x) for x in ids.view(-1).tolist()]
        else:
            as_list = [int(x) for x in ids]

        # Декодируем через токенизатор HF
        result = self.tokenizer.decode(
            as_list, skip_special_tokens=skip_special_tokens
        )
        
        print(f"[decode] Результат: {result!r}")
        return result

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-chat-template-i-soobshcheniya" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">🧰</span> Chat template и сообщения
</h2>

Многие современные модели используют чат‑шаблоны для структурирования диалога (`system`/`user`/`assistant`). Адаптер делегирует форматирование токенизатору через `apply_chat_template`, избегая ручной сборки строк и несоответствий.

<div style="background: #f0fff4; border-left: 4px solid #48bb78; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #276749; margin-bottom: 6px; font-size: 0.95em;">✅ Результат</div>
  <div style="line-height: 1.7;">Используя <code>apply_chat_template</code>, вы получаете ровно ту разметку, под которую обучалась модель (включая спецтокены и разделители).</div>
</div>

<div style="background: #1a202c; color: #a0aec0; padding: 8px 16px; border-radius: 10px 10px 0 0; margin-bottom: -15px; font-family: 'Fira Code', monospace; font-size: 0.8em; display: flex; justify-content: space-between; align-items: center;">
  <span>🐍 nano-infer/python/infer/tokenizer/tokenize.py</span>
  <span style="background: #2d3748; padding: 2px 10px; border-radius: 10px; font-size: 0.85em;">Python</span>
</div>

In [10]:
# Демонстрация форматирования диалога через chat template
def demo_build_prompt() -> str:
    """
    Description:
    ---------------
        Строит текст подсказки из сообщений, используя адаптер.

    Args:
    ---------------
        None

    Returns:
    ---------------
        Готовая строка подсказки для кодирования/генерации.

    Raises:
    ---------------
        Нет явных исключений в этом примере.

    Examples:
    ---------------
        >>> s = demo_build_prompt()
        >>> isinstance(s, str)
        True
    """
    # Создаём адаптер с моделью по умолчанию
    adapter = TokenizerAdapter(model_id=DEFAULT_MODEL_ID)
    # Формируем список сообщений в формате role/content
    messages = [
        {"role": "system", "content": "Вы — полезный ассистент."},
        {"role": "user", "content": "Объясни токенизацию простыми словами."},
    ]
    # Генерируем строку подсказки через chat template
    prompt = adapter.apply_chat_template(messages)
    # Возвращаем строку — её можно передать в encode или генератор
    return prompt

In [13]:
demo_build_prompt()

[TokenizerAdapter] Итог: eos_token_id=151645, pad_token_id=151643
[TokenizerAdapter] Размер словаря: 151,643 токенов


'<|im_start|>system\nВы — полезный ассистент.<|im_end|>\n<|im_start|>user\nОбъясни токенизацию простыми словами.<|im_end|>\n<|im_start|>assistant\n'

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-batch-tokenizaciya" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">📊</span> Батч‑токенизация
</h2>

<div style="display: flex; gap: 16px; margin: 16px 0; align-items: flex-start;">
  <div style="width: 36px; height: 36px; flex: 0 0 36px; flex-shrink: 0; background: linear-gradient(135deg, #667eea, #764ba2); border-radius: 50%; display: flex; align-items: center; justify-content: center; color: white; font-weight: 700; font-size: 0.9em; box-shadow: 0 3px 10px rgba(102,126,234,0.3);">1</div>
  <div style="flex: 1; padding-top: 6px; color: #ffffff; line-height: 1.7;">
    <strong>Соберите список входных строк</strong><br>
    Убедитесь, что список не пуст и содержит строки разумной длины.
  </div>
</div>
<div style="display: flex; gap: 16px; margin: 16px 0; align-items: flex-start;">
  <div style="width: 36px; height: 36px; flex: 0 0 36px; flex-shrink: 0; background: linear-gradient(135deg, #667eea, #764ba2); border-radius: 50%; display: flex; align-items: center; justify-content: center; color: white; font-weight: 700; font-size: 0.9em; box-shadow: 0 3px 10px rgba(102,126,234,0.3);">2</div>
  <div style="flex: 1; padding-top: 6px; color: #ffffff; line-height: 1.7;">
    <strong>Определите max_length</strong><br>
    Выберите целевую длину для паддинга/усечения, чтобы стабилизировать размер тензоров и память.
  </div>
</div>
<div style="display: flex; gap: 16px; margin: 16px 0; align-items: flex-start;">
  <div style="width: 36px; height: 36px; flex: 0 0 36px; flex-shrink: 0; background: linear-gradient(135deg, #667eea, #764ba2); border-radius: 50%; display: flex; align-items: center; justify-content: center; color: white; font-weight: 700; font-size: 0.9em; box-shadow: 0 3px 10px rgba(102,126,234,0.3);">3</div>
  <div style="flex: 1; padding-top: 6px; color: #ffffff; line-height: 1.7;">
    <strong>Вызовите batch_encode</strong><br>
    Передайте список, <code>padding=True</code>, <code>truncation=True</code> и <code>max_length</code> для стабильных батчей.
  </div>
</div>

<div style="background: #1a202c; color: #a0aec0; padding: 8px 16px; border-radius: 10px 10px 0 0; margin-bottom: -15px; font-family: 'Fira Code', monospace; font-size: 0.8em; display: flex; justify-content: space-between; align-items: center;">
  <span>🐍 nano-infer/python/infer/tokenizer/tokenize.py</span>
  <span style="background: #2d3748; padding: 2px 10px; border-radius: 10px; font-size: 0.85em;">Python</span>
</div>

In [19]:
# Демонстрация батч‑кодирования с паддингом до max_length
def demo_batch_encode(max_length: int = 36) -> Dict[str, torch.Tensor]:
    """
    Description:
    ---------------
        Кодирует список строк в батарее примеров и возвращает
        словарь тензоров с одинаковой длиной.

    Args:
    ---------------
        max_length: Целевая длина последовательности после паддинга.

    Returns:
    ---------------
        Dict[str, torch.Tensor]: Поля 'input_ids', 'attention_mask'.

    Raises:
    ---------------
        ValueError: Если внутренняя валидация входных данных не проходит.

    Examples:
    ---------------
        >>> out = demo_batch_encode(max_length=16)
        >>> out["input_ids"].shape[1]
        16
    """
    # Инициализируем адаптер токенизатора
    adapter = TokenizerAdapter(model_id=DEFAULT_MODEL_ID)
    # Готовим список входных строк
    examples: List[str] = [
        "Привет!",                                            # Короткая строка
        "Расскажите, пожалуйста, как работает токенизация?",  # Длиннее
        "OK",                                                 # Очень короткая
    ]
    # Вызываем batch_encode с паддингом до max_length
    batch = adapter.batch_encode(
        examples,
        max_length=max_length,
        padding=True,
        truncation=True,
    )
    # Возвращаем тензоры для дальнейшей обработки
    return batch

In [20]:
demo_batch_encode()

[TokenizerAdapter] Итог: eos_token_id=151645, pad_token_id=151643
[TokenizerAdapter] Размер словаря: 151,643 токенов
[batch_encode] input_ids shape: torch.Size([3, 36]), dtype=torch.int32
[batch_encode] PAD токенов на строку: [32, 17, 35]


{'input_ids': tensor([[ 53645,  26991,   8178,      0, 151643, 151643, 151643, 151643, 151643,
          151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643,
          151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643,
          151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643],
         [ 33504,  28400, 143847,  25965,     11,   5063,  21259,  15952,  82580,
          126309,     11,  51670, 136533,  10813,  14746,   9684,   8959,  55760,
              30, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643,
          151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643],
         [  3925, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643,
          151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643,
          151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643,
          151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643]],

![](/Users/me/Documents/Inference/tutorials/Lesson-01/Infografika/01.png)


**Три строки — три разные судьбы**

`"OK"` — всего 1 токен. `"Привет!"` — 4 токена. `"Расскажите, пожалуйста, как работает токенизация?"` — 19 токенов. Токенизатор Qwen3 работает на субсловных единицах (BPE), поэтому русское слово может разбиться на 2–5 кусков: `"Расскажите"` → `["▁Рас", "ска", "жите"]`.

---

**Зачем нужен PAD токен**

Матрица тензора прямоугольная — все строки обязаны иметь одинаковую длину. Чтобы привести их к единой длине `max_length=36`, токенизатор добавляет специальный токен-заглушку `<pad>` с `id=151643`. У `"OK"` таких заглушек 35, у длинной строки — 17.

---

**`attention_mask` — это фильтр**

Здесь и нужна маска. Модель Transformer обрабатывает весь батч целиком, включая паддинг. Без маски она бы «смотрела» на заглушки так же, как на реальные слова, и усредняла бы смысл с мусором.

- `1` → этот токен настоящий, учитывай его при вычислении внимания
- `0` → это паддинг, не трать на него внимание

На практике модель перед softmax в слое attention добавляет к позициям с маской `0` очень большое отрицательное число (`-inf`), что делает их вес после softmax практически нулевым.

---

**Почему `pad_token_id = 151643`, а `vocab_size = 151 643`?**

Обратите внимание на вывод `__init__`: размер словаря равен 151 643, а `pad_token_id = 151 643` — это индекс за пределами словаря (он равен `vocab_size`, то есть `vocab_size + 0`, если считать с нуля). Это намеренный выбор Qwen3: выделить для PAD отдельный слот вне основного словаря, чтобы ни один реальный токен с ним не совпадал. В логах `__init__` видно, что `eos_token_id=151645` и `pad_token_id=151643` — оба выставлены явно в конфиге модели, поэтому фолбэк `pad = eos` не сработал.

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-dekodirovanie-i-striming" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">🔄</span> Декодирование и стриминг
</h2>

Декодирование — обратное преобразование индексов токенов в текст. При потоковой генерации полезно выводить промежуточные фрагменты без разрывов в середине юникод‑графем или байт‑пар.

<div style="background: #fffbeb; border-left: 4px solid #f59e0b; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #b7791f; margin-bottom: 6px; font-size: 0.95em;">💡 Совет</div>
  <div style="line-height: 1.7;">Для стрима добавляйте минимальные эвристики: декодируйте при появлении пробелов/знаков препинания или каждые N токенов.</div>
</div>

<div style="background: #fff5f5; border-left: 4px solid #fc8181; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #c53030; margin-bottom: 6px; font-size: 0.95em;">⚠️ Внимание</div>
  <div style="line-height: 1.7;">Не декодируйте на каждом токене: это может быть дорого и рвано визуально.</div>
</div>

<div style="background: #1a202c; color: #a0aec0; padding: 8px 16px; border-radius: 10px 10px 0 0; margin-bottom: -15px; font-family: 'Fira Code', monospace; font-size: 0.8em; display: flex; justify-content: space-between; align-items: center;">
  <span>🐍 nano-infer/python/infer/tokenizer/detokenize.py</span>
  <span style="background: #2d3748; padding: 2px 10px; border-radius: 10px; font-size: 0.85em;">Python</span>
</div>

In [ ]:
# Простой стрим‑декодер и проверка на round‑trip

_SAFE_CHARS: set[str] = {" ", ".", ",", "!", "?", ";", ":", "\n"}


def _is_boundary_token(adapter: TokenizerAdapter, token_id: int) -> bool:
    """
    Description:
    ---------------
        Проверяет, является ли одиночный токен безопасной точкой разрыва.
        Декодирует ровно один токен без skip_special_tokens, чтобы не
        потерять пробельные токены (они тоже спецсимволы у части моделей).
    Args:
    ---------------
        adapter   : Инициализированный TokenizerAdapter.
        token_id  : ID токена для проверки.
    Returns:
    ---------------
        True, если декодированный текст токена состоит только из символов
        из _SAFE_CHARS (например: ' ', '.', ',').
    """
    decoded = adapter.tokenizer.decode([token_id], skip_special_tokens=False)
    return bool(decoded) and all(ch in _SAFE_CHARS for ch in decoded)


def safe_stream_decode(
    adapter: TokenizerAdapter,
    buffer: List[int],
    *,
    force_every: int = 8,
) -> Tuple[str, List[int]]:
    """
    Description:
    ---------------
        Декодирует накопленные токены, если появился токен-разделитель
        (пробел/пунктуация) или накопилось force_every токенов.

        Разбиение происходит по индексу токена в буфере, а не по символьной
        позиции в строке — это исключает потерю токенов при BPE-декодировании.
    Args:
    ---------------
        adapter    : Экземпляр TokenizerAdapter для декодирования.
        buffer     : Список накопленных токенов.
        force_every: Принудительная частота декодирования по числу токенов.
    Returns:
    ---------------
        Кортеж (фрагмент текста для вывода, оставшийся буфер токенов).
        Если ещё рано декодировать — возвращает ("", buffer) без изменений.
    Raises:
    ---------------
        ValueError: Если передан неинициализированный адаптер или буфер.
    Examples:
    ---------------
        >>> adapter = TokenizerAdapter()
        >>> text, rest = safe_stream_decode(adapter, [1, 2, 3])
        >>> isinstance(text, str)
        True
    """
    if adapter is None:
        raise ValueError("Адаптер не должен быть None")
    if buffer is None:
        raise ValueError("Буфер токенов не должен быть None")
    if not buffer:
        return "", []

    print(f"[safe_stream_decode] буфер={buffer} (len={len(buffer)}, force_every={force_every})")

    # ── Ветка 1: принудительный сброс по размеру буфера ─────────────────────
    if len(buffer) >= force_every:
        text = adapter.decode(buffer)
        print(f"[safe_stream_decode] force flush → {text!r}")
        return text, []

    # ── Ветка 2: ищем последний токен-разделитель в буфере ──────────────────
    # Идём с конца, чтобы декодировать максимально длинный готовый фрагмент
    last_boundary: int = -1
    for i in range(len(buffer) - 1, -1, -1):
        if _is_boundary_token(adapter, buffer[i]):
            last_boundary = i
            break

    if last_boundary >= 0:
        to_decode = buffer[: last_boundary + 1]   # токены до разделителя включительно
        remaining = buffer[last_boundary + 1 :]   # токены после границы
        text = adapter.decode(to_decode)
        print(f"[safe_stream_decode] boundary at idx={last_boundary} → {text!r}, "
              f"остаток={remaining}")
        return text, remaining

    # ── Ветка 3: разделителя нет, буфер мал — ждём ──────────────────────────
    print("[safe_stream_decode] ждём ещё токенов...")
    return "", buffer


def demo_safe_stream_decode() -> None:
    """
    Description:
    ---------------
        Демонстрирует работу safe_stream_decode в трёх сценариях:
        1. force_every — принудительный сброс по порогу токенов.
        2. Разделитель найден — декодируем до него.
        3. Разделитель не найден — буфер остаётся нетронутым.
        4. Имитация реального стрима — токены приходят по одному.
    """
    adapter = TokenizerAdapter(model_id=DEFAULT_MODEL_ID)

    test_phrase = "Привет, мир. Как дела?"
    all_ids: List[int] = adapter.encode(test_phrase).tolist()

    print(f"\n{'='*60}")
    print(f"Тестовая фраза : {test_phrase!r}")
    print(f"Все токены     : {all_ids}")
    print(f"Всего токенов  : {len(all_ids)}")
    print(f"{'='*60}")

    # ── Сценарий 1: buffer >= force_every ───────────────────────────────────
    print("\n[ Сценарий 1 ] force_every=4, буфер = первые 4 токена")
    buffer_1: List[int] = all_ids[:4]
    print(f"  Буфер до вызова : {buffer_1}")
    text_1, rest_1 = safe_stream_decode(adapter, buffer_1, force_every=4)
    print(f"  Декодировано    : {text_1!r}")
    print(f"  Остаток буфера  : {rest_1}  (пуст — буфер сброшен)")

    # ── Сценарий 2: разделитель найден ──────────────────────────────────────
    print("\n[ Сценарий 2 ] force_every=20, буфер = первые 6 токенов")
    buffer_2: List[int] = all_ids[:6]
    print(f"  Буфер до вызова : {buffer_2}")
    print(f"  Промежуточный текст : {adapter.decode(buffer_2)!r}")
    text_2, rest_2 = safe_stream_decode(adapter, buffer_2, force_every=20)
    print(f"  Декодировано    : {text_2!r}")
    print(f"  Остаток буфера  : {rest_2}")

    # ── Сценарий 3: разделитель не найден, буфер копится ────────────────────
    print("\n[ Сценарий 3 ] force_every=20, буфер = первые 2 токена (нет разделителя)")
    buffer_3: List[int] = all_ids[:2]
    print(f"  Буфер до вызова : {buffer_3}")
    print(f"  Промежуточный текст : {adapter.decode(buffer_3)!r}")
    text_3, rest_3 = safe_stream_decode(adapter, buffer_3, force_every=20)
    print(f"  Декодировано    : {text_3!r}  (пусто — ждём ещё токенов)")
    print(f"  Остаток буфера  : {rest_3}  (буфер сохранён)")

    # ── Сценарий 4: реальный стрим — токены по одному ───────────────────────
    print(f"\n{'='*60}")
    print("[ Стрим ] Токены приходят по одному, force_every=5")
    print(f"{'='*60}")

    stream_buffer: List[int] = []
    output_chunks: List[str] = []

    for i, token_id in enumerate(all_ids):
        stream_buffer.append(token_id)
        print(f"  +токен[{i}]={token_id:>7}  буфер={stream_buffer}")

        chunk, stream_buffer = safe_stream_decode(
            adapter, stream_buffer, force_every=5
        )

        if chunk:
            output_chunks.append(chunk)
            print(f"  → выдали чанк   : {chunk!r}")
        else:
            print(f"  → копим (нет разделителя, буфер < 5)")

    # Хвост: принудительно декодируем всё, что осталось в буфере
    if stream_buffer:
        tail, _ = safe_stream_decode(adapter, stream_buffer, force_every=1)
        if tail:
            output_chunks.append(tail)
            print(f"\n  → хвост буфера  : {tail!r}")

    print(f"\nИтоговый текст: {''.join(output_chunks)!r}")

In [32]:
demo_safe_stream_decode()

[TokenizerAdapter] Итог: eos_token_id=151645, pad_token_id=151643
[TokenizerAdapter] Размер словаря: 151,643 токенов
[encode] Результат:  torch.Size([9]) токенов, dtype=torch.int32
[encode] ID токенов: [53645, 26991, 8178, 11, 137144, 13, 128654, 129691, 30]
[encode] Токены:     ['ÐŁÑĢ', 'ив', 'еÑĤ', ',', 'ĠмиÑĢ', '.', 'ĠÐļак', 'Ġдела', '?']

Тестовая фраза : 'Привет, мир. Как дела?'
Все токены     : [53645, 26991, 8178, 11, 137144, 13, 128654, 129691, 30]
Всего токенов  : 9

[ Сценарий 1 ] force_every=4, буфер = первые 4 токена
  Буфер до вызова : [53645, 26991, 8178, 11]
[safe_stream_decode] буфер=[53645, 26991, 8178, 11] (len=4, force_every=4)
[decode] Результат: 'Привет,'
[safe_stream_decode] force flush → 'Привет,'
  Декодировано    : 'Привет,'
  Остаток буфера  : []  (пуст — буфер сброшен)

[ Сценарий 2 ] force_every=20, буфер = первые 6 токенов
  Буфер до вызова : [53645, 26991, 8178, 11, 137144, 13]
[decode] Результат: 'Привет, мир.'
  Промежуточный текст : 'Привет, 

# Разбор вывода `demo_safe_stream_decode`

## Что происходит в самом начале

```
[TokenizerAdapter] Итог: eos_token_id=151645, pad_token_id=151643
[TokenizerAdapter] Размер словаря: 151,643 токенов
```

Перед любой работой со стримом адаптер загружает токенизатор Qwen3 и фиксирует два служебных токена. Они не несут смысловой нагрузки, но нужны модели как технические маркеры:

- `eos_token_id=151645` — сигнал конца генерации («стоп, текст закончился»)
- `pad_token_id=151643` — заглушка для выравнивания длин в батче

---

## Кодирование фразы

```
[encode] Результат:  torch.Size([9]) токенов, dtype=torch.int32
[encode] ID токенов: [53645, 26991, 8178, 11, 137144, 13, 128654, 129691, 30]
[encode] Токены:     ['ÐŁÑĢ', 'ив', 'еÑĤ', ',', 'ĠмиÑĢ', '.', 'ĠÐļак', 'Ġдела', '?']
```

Фраза `'Привет, мир. Как дела?'` разбилась на **9 токенов**. Сырые названия токенов выглядят как кракозябры — это нормально: `convert_ids_to_tokens` возвращает внутреннее BPE-представление в байтовой нотации, а не читаемый текст. Реальные значения видны при декодировании через `adapter.decode()`.

Важное наблюдение: русские слова разбиваются на несколько кусков, а знаки препинания получают **отдельный токен** — именно по ним `safe_stream_decode` будет находить границы:

| ID | Что декодирует |
|---|---|
| 53645, 26991, 8178 | `При`, `вет` — слово целиком |
| **11** | `,` — **разделитель** |
| 137144 | ` мир` — с пробелом внутри |
| **13** | `.` — **разделитель** |
| 128654, 129691 | ` Как`, ` дела` |
| **30** | `?` — **разделитель** |

---

## Сценарий 1 — принудительный сброс по `force_every`

```
[ Сценарий 1 ] force_every=4, буфер = первые 4 токена
  Буфер до вызова : [53645, 26991, 8178, 11]
[safe_stream_decode] буфер=[53645, 26991, 8178, 11] (len=4, force_every=4)
[safe_stream_decode] force flush → 'Привет,'
  Декодировано    : 'Привет,'
  Остаток буфера  : []  (пуст — буфер сброшен)
```

Буфер достиг порога `force_every=4` — функция не ищет разделители, а немедленно декодирует **всё** и возвращает пустой остаток. Это защитный механизм: если модель долго генерирует длинное слово без знаков препинания, пользователь всё равно увидит текст, а не замороженный экран.

> **Когда срабатывает:** `len(buffer) >= force_every` — первая проверка в функции, самая приоритетная.

---

## Сценарий 2 — разделитель найден

```
[ Сценарий 2 ] force_every=20, буфер = первые 6 токенов
  Буфер до вызова : [53645, 26991, 8178, 11, 137144, 13]
[safe_stream_decode] boundary at idx=5 → 'Привет, мир.', остаток=[]
  Декодировано    : 'Привет, мир.'
  Остаток буфера  : []
```

Порог `force_every=20` не достигнут, поэтому функция ищет **последний токен-разделитель** с конца буфера. Токен с `idx=5` — это `.` (`id=13`), он является чистым разделителем. Функция декодирует токены `[0..5]` включительно — получает `'Привет, мир.'` — и возвращает пустой остаток, потому что разделитель стоял в самом конце буфера.

> **Ключевой момент:** поиск идёт **с конца**, чтобы захватить максимально длинный готовый фрагмент.

---

## Сценарий 3 — разделителя нет, ждём

```
[ Сценарий 3 ] force_every=20, буфер = первые 2 токена
  Буфер до вызова : [53645, 26991]
[safe_stream_decode] ждём ещё токенов...
  Декодировано    : ''  (пусто — ждём ещё токенов)
  Остаток буфера  : [53645, 26991]  (буфер сохранён)
```

Токены `53645` и `26991` декодируются в `'Прив'` — это середина слова, никаких разделителей нет, порог не достигнут. Функция возвращает `("", buffer)` — **буфер остаётся нетронутым**. Это правильное поведение: выводить неполное слово нельзя, нужно дождаться следующих токенов.

---

## Стрим — токены приходят по одному

Это главный сценарий использования. Разберём каждый шаг:

```
+токен[0]=53645  → копим   (буфер=['При...'])
+токен[1]=26991  → копим   (буфер=['При...', 'вет'])
+токен[2]=8178   → копим   (буфер=['При...', 'вет', 'ет'])
+токен[3]=11     → выдаём 'Привет,'   остаток=[]
```

Запятая `,` (`id=11`) — первый разделитель. Функция декодирует накопленные 4 токена и сбрасывает буфер. Пользователь видит `'Привет,'`.

```
+токен[4]=137144  → копим   (буфер=[' мир'])
+токен[5]=13      → выдаём ' мир.'   остаток=[]
```

Точка `.` (`id=13`) — следующий разделитель. Буфер был всего 2 токена, но разделитель найден — выдаём сразу. Пользователь видит `' мир.'`.

```
+токен[6]=128654  → копим   (буфер=[' Как'])
+токен[7]=129691  → копим   (буфер=[' Как', ' дела'])
+токен[8]=30      → выдаём ' Как дела?'   остаток=[]
```

Вопросительный знак `?` (`id=30`) завершает последний фрагмент. Два слова `' Как'` и `' дела'` накопились вместе и вышли единым чанком.

---

## Итог

```
Итоговый текст: 'Привет, мир. Как дела?'
```

Все три чанка склеиваются в исходную фразу без потерь.

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-diagnostika-i-otladka" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">🔧</span> Диагностика и отладка
</h2>

<div style="background: #fffbeb; border-left: 4px solid #f59e0b; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #b7791f; margin-bottom: 6px; font-size: 0.95em;">💡 Совет</div>
  <div style="line-height: 1.7;">Если <code>pad_token_id</code> не определён — выставляйте его равным <code>eos_token_id</code>, это типовой приём для чат‑моделей.</div>
</div>

<div style="background: #fff5f5; border-left: 4px solid #fc8181; padding: 16px 20px; border-radius: 0 10px 10px 0; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #c53030; margin-bottom: 6px; font-size: 0.95em;">⚠️ Внимание</div>
  <div style="line-height: 1.7;">Следите за типами: некоторые рантаймы ожидают <code>int32</code>. Несогласованность типов может приводить к лишним копиям/ошибкам.</div>
</div>

<div style="background: #f8f9ff; border: 1px dashed #667eea; padding: 16px 20px; border-radius: 10px; margin: 12px 0; color: #2d3748;">
  <div style="font-weight: 700; color: #667eea; margin-bottom: 6px; font-size: 0.95em;">🔗 См. также</div>
  <div style="line-height: 1.7;">Проверьте <code>tokenizer.backend_tokenizer</code> для fast‑версий — полезно для глубокой диагностики.</div>
</div>

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-checklist-dod" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">✅</span> Чеклист DoD
</h2>

<table style="width: 100%; border-collapse: collapse; margin: 20px 0; border-radius: 12px; overflow: hidden; box-shadow: 0 2px 10px rgba(0,0,0,0.06);">
  <thead>
    <tr style="background: linear-gradient(135deg, #667eea, #764ba2);">
      <th style="padding: 14px 18px; text-align: left; color: white; font-weight: 600; font-size: 0.95em;">Пункт</th>
      <th style="padding: 14px 18px; text-align: left; color: white; font-weight: 600; font-size: 0.95em;">Статус</th>
    </tr>
  </thead>
  <tbody>
    <tr style="background: #f8f9ff;">
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Единый API (encode/batch/decode/chat)</td>
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Готово</td>
    </tr>
    <tr style="background: #ffffff;">
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Типы тензоров int32</td>
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Готово</td>
    </tr>
    <tr style="background: #f8f9ff;">
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Корректные pad/eos</td>
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Готово</td>
    </tr>
    <tr style="background: #ffffff;">
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Тесты на round‑trip и батч</td>
      <td style="padding: 12px 18px; border-bottom: 1px solid #e2e8f0; color: #2d3748;">Готово</td>
    </tr>
  </tbody>
</table>

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-ssylki-i-resursy" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">🔗</span> Ссылки и ресурсы
</h2>

- Hugging Face Tokenizers API: https://huggingface.co/docs/transformers/main_classes/tokenizer
- Chat templates: https://huggingface.co/docs/transformers/chat_templating
- Модель Qwen3: https://huggingface.co/Qwen
- PyTorch: https://pytorch.org/

<div style="text-align: center; margin: 30px 0;">
  <div style="display: inline-block; width: 60px; height: 3px; background: linear-gradient(to right, transparent, #667eea, transparent); border-radius: 2px;"></div>
</div>

<h2 id="section-voprosy-dlya-samoproverki" style="color: #ffffff; font-size: 1.8em; font-weight: 700; border-bottom: 3px solid #667eea; padding-bottom: 12px; margin-top: 40px; margin-bottom: 20px; letter-spacing: -0.01em;">
  <span style="background: linear-gradient(135deg, #667eea, #764ba2); -webkit-background-clip: text; -webkit-text-fill-color: transparent; margin-right: 8px;">❓</span> Вопросы для самопроверки
</h2>

<div style="background: #fffff0; border: 1px solid #ecc94b40; border-radius: 14px; padding: 24px 28px; margin: 20px 0;">
  <h4 style="color: #975a16; margin: 0 0 14px 0; font-size: 1.1em; font-weight: 700;">🤔 Вопросы для самопроверки</h4>
  <ol style="color: #2d3748; line-height: 2; margin: 0; padding-left: 20px;">
    <li>Почему важно выставлять <code>pad_token_id</code>, если он отсутствует?</li>
    <li>В чём преимущества использования <code>apply_chat_template</code>?</li>
    <li>Как выбор <code>max_length</code> влияет на память и время?</li>
  </ol>
</div>

<div style="background: linear-gradient(135deg, rgba(30, 41, 59, 0.9), rgba(45, 55, 72, 0.9)); border: 1px solid rgba(102,126,234,0.5); border-radius: 16px; padding: 30px; margin-top: 30px;">
  <!-- Заголовок стал белым -->
  <h3 style="color: #ffffff; margin: 0 0 18px 0; font-size: 1.3em; font-weight: 700;">🎯 Ключевые выводы</h3>
  
  <!-- Текст списка стал белым -->
  <ul style="color: #ffffff; line-height: 2; list-style: none; padding: 0; margin: 0;">
    <!-- Границы между пунктами сделал чуть ярче для темного фона -->
    <li style="padding: 6px 0; border-bottom: 1px solid rgba(255,255,255,0.1);">✅ Единый адаптер упрощает работу с токенизаторами HF и стабилизирует рантайм.</li>
    <li style="padding: 6px 0; border-bottom: 1px solid rgba(255,255,255,0.1);">✅ Корректные <code style="background: rgba(255,255,255,0.1); padding: 2px 6px; border-radius: 4px; color: #fff;">pad/eos</code> и тип <code style="background: rgba(255,255,255,0.1); padding: 2px 6px; border-radius: 4px; color: #fff;">int32</code> критичны для совместимости.</li>
    <li style="padding: 6px 0;">✅ Тесты на round‑trip и батч помогают быстро ловить регрессии.</li>
  </ul>
</div>

<div style="text-align: center; padding: 30px 20px; margin-top: 40px; border-top: 2px solid #e2e8f0; color: #718096; font-size: 0.85em;">
  <p style="margin: 0 0 8px 0;">📘 <strong style="color: #667eea;">CookBook Series</strong> — Lesson 01 — Токенизатор и адаптер Hugging Face (Qwen3)</p>
  <p style="margin: 0;">Обратная связь приветствуется</p>
</div>